### PR TITLE
docs: added new features in the monitoring document & modified the docs that had minor changes

### PR DESCRIPTION
**Pull request by balaharish7**
- Added new features in the monitoring document
- Reviewed and modified the docs that had minor changes
- Modified the Upgrade document with the latest changes

### CODE REVIEW

File: docs/main/introduction-to-openebs/features.mdx [hunk 1]
```diff
@@ -65,7 +65,7 @@ OpenEBS Features, like any storage solution, can be broadly classified into the
 
 <TwoColumn left="1fr" right="200px">
   <p>
-    The backup and restore of OpenEBS volumes works with Kubernetes backup and restore solutions such as Velero (formerly Heptio Ark) via open source OpenEBS Velero-plugins. Data backup to object storage targets such as AWS S3, GCP Object Storage or MinIO are frequently deployed using the OpenEBS incremental snapshot capability. This storage level snapshot and backup saves a significant amount of bandwidth and storage space as only incremental data is used for backup.
+    The backup and restore of OpenEBS volumes works with Kubernetes backup and restore solutions such as Velero (f.k.a Heptio Ark) via open source OpenEBS Velero-plugins. Data backup to object storage targets such as AWS S3, GCP Object Storage or MinIO are frequently deployed using the OpenEBS incremental snapshot capability. This storage level snapshot and backup saves a significant amount of bandwidth and storage space as only incremental data is used for backup.
```
Comment (avishnu) [inline docs/main/introduction-to-openebs/features.mdx:68]: we can remove f.k.a Heptio Ark, that's very long time ago and now, Velero name has established it's popularity

File: docs/main/introduction-to-openebs/introduction-to-openebs.md [hunk 2]
```diff
@@ -22,7 +22,7 @@ The [OpenEBS Adoption stories](https://github.com/openebs/openebs/blob/master/AD
 
 - OpenEBS provides consistency across all Kubernetes distributions - On-premise and Cloud.
 - OpenEBS with Kubernetes increases Developer and Platform SRE Productivity.
-- OpenEBS is Easy to use compared to other solutions, for eg trivial to install & enabling entirely dynamic provisioning.
+- OpenEBS is Easy to use compared to other solutions, for eg trivial to install and enabling entirely dynamic provisioning.
```
Comment (avishnu) [inline docs/main/introduction-to-openebs/introduction-to-openebs.md:25]: OpenEBS scores in its ease of use over other solutions. It is trivial to setup, install and configure.

File: docs/main/troubleshooting/troubleshooting-local-storage.md [hunk 3]
```diff
@@ -29,26 +29,26 @@ kubectl get bdc -n openebs
 shows stale `Pending` BDCs created by localpv provisioner, even after the corresponding PVC has been deleted.
```
Comment (avishnu) [inline docs/main/troubleshooting/troubleshooting-local-storage.md:29]: BDCs are related to NDM, need to remove those contents.

File: docs/main/troubleshooting/troubleshooting-local-storage.md [hunk 4]
```diff
@@ -62,16 +62,15 @@ The following are different types of messages shown when the node on which local
...
 
 2. No devices with matching criteria
 
 ```shell hideCopy
 Warning  SelectionFailed  6m25s (x18 over 11m)  blockdeviceclaim-operator  no devices found matching the criteria
 ```
 
-It means that the there are no devices for claiming after filtering based on filesystem type and node name. Make sure the blockdevices on the node
-have the correct filesystem as mentioned in the storage class (default is `ext4`)
+It means that the there are no devices for claiming after filtering based on filesystem type and node name. Make sure the blockdevices on the node have the correct filesystem as mentioned in the storage class (default is `ext4`).
```
Comment (avishnu) [inline docs/main/troubleshooting/troubleshooting-local-storage.md:74]: Many contents seem to be outdated (belong to deprecated engine) in this page. You can remove them in a separate PR.

File: docs/main/user-guides/replicated-storage-user-guide/advanced-operations/monitoring.md [hunk 5]
```diff
@@ -101,4 +101,107 @@ Upon successful integration of the exporter with the Prometheus stack, the metri
 | kubelet_volume_stats_inodes_used | Gauge | Integer | The total number of inodes that have been utilized to store metadata. |
 
 
+## Performance Monitoring Stack
+
+Earlier, the pool capacity/state stats were exported, and the exporter used to cache the metrics and return when Prometheus client queried. This was not ensuring the latest data retuns during the Prometheus poll cycle.
+
+This feature exports performace stats for Volume, Replica and DiskPool resource. Mayastor Prometheus exporter runs along with IO engine container and exposes Prometheus counters. In addition to that, exporter does not pre-populate the metrics. It queries the IO engine, updates cache, and returns the latest metrics when the Prometheus client queries the exporter.
+
+With this change, the exporter calls to the IO engine for stats tied to the frequency of Prometheus poll cycle.
```
Comment (avishnu) [inline docs/main/user-guides/replicated-storage-user-guide/advanced-operations/monitoring.md:110]: Can this sentence be grammatically improved?

Comment (balaharish7) [reply docs/main/user-guides/replicated-storage-user-guide/advanced-operations/monitoring.md:110]: Is the below sentence fine @avishnu 

"With this change, the exporter will query the IO engine for stats related to the frequency of Prometheus poll cycle."

File: docs/main/user-guides/upgrades.md [hunk 6]
```diff
@@ -73,6 +73,10 @@ This section describes the Replicated Storage upgrade from OpenEBS Umbrella char
 
 1. Start the helm upgrade process with the new chart, i.e. 4.0.0 by using the below command:
 
+:::caution
+For users who needs hot upgrade, `--set mayastor.agents.core.rebuild.partial.enabled=false` is mandatory while upgrading to OpenEBS 4.0.0 to ensure there are no issues in the upgrade process. For users who can scale down the applications in the upgrade process can skip using this flag.
```
Comment (avishnu) [inline docs/main/user-guides/upgrades.md:77]: It is highly recommended to disable partial rebuild during upgrade to OpenEBS 4.0.0 to ensure data consistency during upgrade. Pass the value `..` in the upgrade command.

File: docs/main/user-guides/upgrades.md [hunk 7]
```diff
@@ -126,8 +131,19 @@ openebs-upgrade-v2-6-0-s58xl                   0/1     Completed   0          7m
 
 4. Once the upgrade process completes, all the volumes and pools should be online.
 
+5. Users who had opted for hot upgrade and disabled the partial rebuild using `--set mayastor.agents.core.rebuild.partial.enabled=true`, can enable it back with the below command.
```
Comment (avishnu) [inline docs/main/user-guides/upgrades.md:134]: Modify the sentence as per the above comment

File: docs/main/user-guides/upgrades.md [hunk 8]
```diff
@@ -126,8 +131,19 @@ openebs-upgrade-v2-6-0-s58xl                   0/1     Completed   0          7m
...
+
+:::info
+This step is not applicable (or can be skipped) if you have not opted for hot upgrade.
```
Comment (avishnu) [inline docs/main/user-guides/upgrades.md:137]: Here as well.

File: docs/main/user-guides/upgrades.md (hunk 7 above)
Comment (balaharish7) [reply docs/main/user-guides/upgrades.md:134]: We have modified the sentence like this: 

If you have disabled the partial rebuild during upgrade, re-enable it by adding the value `--set mayastor.agents.core.rebuild.partial.enabled=true` in the upgrade command.

Hope this looks fine @avishnu

File: docs/main/user-guides/replicated-storage-user-guide/advanced-operations/monitoring.md (hunk 5 above)
Comment (avishnu) [reply docs/main/user-guides/replicated-storage-user-guide/advanced-operations/monitoring.md:110]: 'the stats related to the frequency..' is not correct.

File: docs/main/user-guides/upgrades.md [hunk 9]
```diff
@@ -131,10 +131,10 @@ openebs-upgrade-v2-6-0-s58xl                   0/1     Completed   0          7m
 
 4. Once the upgrade process completes, all the volumes and pools should be online.
 
-5. Users who had opted for hot upgrade and disabled the partial rebuild using `--set mayastor.agents.core.rebuild.partial.enabled=true`, can enable it back with the below command.
+5. If you have disabled the partial rebuild during the upgrade, re-enable it by adding the value `--set mayastor.agents.core.rebuild.partial.enabled=true` in the upgrade command.
 
 :::info
```
Comment (avishnu) [inline docs/main/user-guides/upgrades.md:136]: let's remove this info note

File: docs/main/user-guides/replicated-storage-user-guide/advanced-operations/monitoring.md [hunk 10]
```diff
@@ -101,4 +101,109 @@ Upon successful integration of the exporter with the Prometheus stack, the metri
 | kubelet_volume_stats_inodes_used | Gauge | Integer | The total number of inodes that have been utilized to store metadata. |
 
 
+## Performance Monitoring Stack
+
+Earlier, the pool capacity/state stats were exported, and the exporter used to cache the metrics and return when Prometheus client queried. This was not ensuring the latest data retuns during the Prometheus poll cycle.
+
+This feature exports performace stats for Volume, Replica and DiskPool resource. Mayastor Prometheus exporter runs along with IO engine container and exposes Prometheus counters. In addition to that, exporter does not pre-populate the metrics. It queries the IO engine, updates cache, and returns the latest metrics when the Prometheus client queries the exporter.
```
Comment (avishnu) [inline docs/main/user-guides/replicated-storage-user-guide/advanced-operations/monitoring.md:108]: performance*

Remove the "Mayastor Prometheus exporter..." as we use the same exporter which previously existed and exported other metrics.

We need a bit of re-formatting here.

In addition to the capacity and state metrics, the metrics exporter also exports performance statistics for pools, volumes and replicas as Prometheus counters. The exporter does not pre-fetch or cache the metrics, it polls the IO engine inline with the Prometheus client pooling cycle.

File: docs/main/user-guides/replicated-storage-user-guide/advanced-operations/monitoring.md [hunk 11]
```diff
@@ -101,4 +101,109 @@ Upon successful integration of the exporter with the Prometheus stack, the metri
...
+
+With this change, the exporter calls to the IO engine for stats tied to the frequency of Prometheus poll cycle.
+
+With this change, the exporter will query the IO engine for stats related to the frequency of Prometheus poll cycle.
```
Comment (avishnu) [inline docs/main/user-guides/replicated-storage-user-guide/advanced-operations/monitoring.md:112]: Remove this line

File: docs/main/user-guides/replicated-storage-user-guide/advanced-operations/monitoring.md [hunk 12]
```diff
@@ -101,4 +101,109 @@ Upon successful integration of the exporter with the Prometheus stack, the metri
...
+read_throughput = bytes_read (current poll) - bytes_read (previous_poll) / poll period (in sec)
+```
+
+```
+write_throughput = bytes_written (current poll) - bytes_written (previous_poll) / poll period (in sec)
+```
+
+### Handling Counter Reset
+
+If a particular IO engine restarts and when the IO engine comes back online, all the bdev stats will be reset. Users will receive lesser values for all the resources residing on that particular IO engine due to reset. So using the above logic would yield negative values. Hence, the counter current poll is less than the counter previous poll. In this case, do the following:
```
Comment (avishnu) [inline docs/main/user-guides/replicated-storage-user-guide/advanced-operations/monitoring.md:195]: The performance stats are not persistent across IO engine restart, this means the counters will be reset upon IO engine restart.

File: docs/main/user-guides/replicated-storage-user-guide/advanced-operations/monitoring.md (hunk 5 above)
Comment (balaharish7) [reply docs/main/user-guides/replicated-storage-user-guide/advanced-operations/monitoring.md:110]: Is it "stats tied to the frequency"? @avishnu

File: docs/main/user-guides/replicated-storage-user-guide/advanced-operations/monitoring.md [hunk 13]
```diff
@@ -105,11 +105,7 @@ Upon successful integration of the exporter with the Prometheus stack, the metri
 
 Earlier, the pool capacity/state stats were exported, and the exporter used to cache the metrics and return when Prometheus client queried. This was not ensuring the latest data retuns during the Prometheus poll cycle.
 
-This feature exports performace stats for Volume, Replica and DiskPool resource. Mayastor Prometheus exporter runs along with IO engine container and exposes Prometheus counters. In addition to that, exporter does not pre-populate the metrics. It queries the IO engine, updates cache, and returns the latest metrics when the Prometheus client queries the exporter.
-
-With this change, the exporter calls to the IO engine for stats tied to the frequency of Prometheus poll cycle.
-
-With this change, the exporter will query the IO engine for stats related to the frequency of Prometheus poll cycle.
+In addition to the capacity and state metrics, the metrics exporter also exports performance statistics for pools, volumes, and replicas as Prometheus counters. The exporter does not pre-fetch or cache the metrics, it polls the IO engine inline with the Prometheus client pooling cycle.
```
Comment (avishnu) [inline docs/main/user-guides/replicated-storage-user-guide/advanced-operations/monitoring.md:108]: polling cycle

File: docs/main/glossary.md [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+| OpenEBS |	Open Elastic Block Store |
+| PV | Persistent Volume |
+| PVC |	Persistent Volume Claim |
+| RBAC | Role-Based Access Control |
+| SPDK | Storage Performance Development Kit |
+| SRE |	Site Reliability Engineering |
+| TCP |	Transmission Control Protocol |
+| VG | Volume Group |
+| YAML | Yet Another Markup Language |
+| ZFS |	Zettabyte File System |
```
Comment (avishnu) [inline docs/main/glossary.md:41]: Good

**Review by avishnu — APPROVED** [review]
LGTM

**Pull request merged**